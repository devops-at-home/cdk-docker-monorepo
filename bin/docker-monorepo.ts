#!/usr/bin/env node
import * as cdk from "@aws-cdk/core";
import { DockerMonorepoPipelineStack } from "../lib/docker-monorepo-pipeline-stack";
import { DockerMonorepoStage } from "../lib/docker-monorepo-stage";

const app = new cdk.App();

new DockerMonorepoPipelineStack(app, "DockerMonorepoPipeline");

// For manual deployments to play account only - run using npm run deploy:play
new DockerMonorepoStage(app, "DockerMonorepoPlay", {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

app.synth();
