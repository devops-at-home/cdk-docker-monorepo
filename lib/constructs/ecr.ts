import * as cdk from "@aws-cdk/core";
import * as ecr from "@aws-cdk/aws-ecr";
import * as iam from "@aws-cdk/aws-iam";

export class ECR extends cdk.Construct {
  public repos: { [key: string]: ecr.Repository } = {};
  private dockerRole: iam.IRole;
  constructor(
    scope: cdk.Construct,
    id: string,
    nameArray: string[],
    dockerRole: iam.IRole
  ) {
    super(scope, id);

    this.dockerRole = dockerRole;

    for (let idx in nameArray) {
      this.__createRepo(nameArray[idx]);
    }
  }
  private __createRepo = (name: string) => {
    this.repos[name] = new ecr.Repository(this, `EcrRepo-${name}`, {
      repositoryName: name,
      imageScanOnPush: true,
      lifecycleRules: [{ maxImageCount: 5 }],
    });
    this.repos[name].grantPullPush(this.dockerRole);
  };
}
