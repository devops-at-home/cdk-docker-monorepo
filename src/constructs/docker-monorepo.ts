import { RemovalPolicy, Aws } from 'aws-cdk-lib';
import {
  Repository,
  RepositoryEncryption,
  TagMutability,
} from 'aws-cdk-lib/aws-ecr';
import { IRole, PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

import { GitHubActionsOidcProvider } from './gh-aws-oidc-connect-provider';
import { GitHubActionsRole } from './gh-aws-oidc-connect-role';

export type DockerMonorepoProps = {
  githubOrg: string;
  githubRepoName: string;
  ecrConfigs: ECRconfig[];
};

type ECRconfig = {
  repositoryName: string;
  maxImageCount: number;
};

export class DockerMonorepo extends Construct {
  public repos: Repository[];
  public role: IRole;
  constructor(scope: Construct, id: string, props: DockerMonorepoProps) {
    super(scope, id);

    const { githubOrg, githubRepoName, ecrConfigs } = props;
    const repository = `${githubOrg}/${githubRepoName}`;
    this.repos = [];

    // Configure OIDC and create the role
    this.role = new GitHubActionsRole(this, 'GitHubActionsRole', {
      provider: GitHubActionsOidcProvider.forAccount(),
      repository,
      inlinePolicies: {
        ecr: new PolicyDocument({
          statements: [
            new PolicyStatement({
              actions: ['ecr:GetAuthorizationToken'],
              resources: ['*'],
            }),
            new PolicyStatement({
              actions: [
                'ecr:ListImages',
                'ecr:BatchCheckLayerAvailability',
                'ecr:BatchGetImage',
                'ecr:CompleteLayerUpload',
                'ecr:InitiateLayerUpload',
                'ecr:PutImage',
                'ecr:UploadLayerPart',
              ],
              // TODO: manually specify repos
              resources: this.getRepoArnsFromConfig(ecrConfigs),
            }),
          ],
        }),
      },
    });

    // Create repos with tags
    for (let config of ecrConfigs) {
      const { repositoryName, maxImageCount } = config;
      const repo = new Repository(this, `Repository-${repositoryName}`, {
        encryption: RepositoryEncryption.AES_256,
        imageScanOnPush: true,
        imageTagMutability: TagMutability.IMMUTABLE,
        repositoryName,
        removalPolicy: RemovalPolicy.DESTROY,
        lifecycleRules: [
          {
            maxImageCount,
          },
        ],
      });
      this.repos.push(repo);
    }
  }
  private getRepoArnsFromConfig = (ecrConfigs: ECRconfig[]): string[] => {
    const { REGION, ACCOUNT_ID } = Aws;
    return ecrConfigs.map((config) => {
      return `arn:aws:ecr:${REGION}:${ACCOUNT_ID}:repository/${config.repositoryName}`;
    });
  };
}
