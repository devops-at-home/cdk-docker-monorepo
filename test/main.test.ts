import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DockerMonorepo } from '../src/constructs/docker-monorepo';
import { dockerMonorepoProps } from '../.projenrc';

test('Snapshot', () => {
  const stack = new Stack();

  new DockerMonorepo(stack, 'DockerMonorepo', dockerMonorepoProps);

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
