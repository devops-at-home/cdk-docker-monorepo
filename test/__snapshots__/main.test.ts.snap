// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DockerMonorepoGitHubActionsRole095A018C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "StringLike": Object {
                  "token.actions.githubusercontent.com:sub": "repo:devops-at-home/*:ref:refs/heads/main*",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "ecr:GetAuthorizationToken",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": Array [
                    "ecr:ListImages",
                    "ecr:BatchCheckLayerAvailability",
                    "ecr:CompleteLayerUpload",
                    "ecr:InitiateLayerUpload",
                    "ecr:PutImage",
                    "ecr:UploadLayerPart",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:ecr:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":repository/hello-kubernetes",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ecr",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DockerMonorepoRepositoryhellokubernetesE94F0461": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ImageScanningConfiguration": Object {
          "ScanOnPush": true,
        },
        "ImageTagMutability": "IMMUTABLE",
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"any\\",\\"countType\\":\\"imageCountMoreThan\\",\\"countNumber\\":10},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "hello-kubernetes",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
